File: app/models/promocode.py
"""Модель промокода."""
import uuid
from datetime import datetime
from decimal import Decimal

from sqlalchemy import String, Numeric, Boolean, ForeignKey, Integer, BigInteger, DateTime, Text
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from app.models.business import Business
    from app.models.order import Order
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Promocode(Base):
    """Модель промокода."""

    __tablename__ = "promocodes"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    business_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("businesses.id"), nullable=False, index=True)
    code: Mapped[str] = mapped_column(String(50), nullable=False, unique=True, index=True)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    
    # Тип скидки: 'percentage' (процент) или 'fixed' (фиксированная сумма)
    discount_type: Mapped[str] = mapped_column(String(20), nullable=False, default="percentage")
    # Значение скидки (процент от 0 до 100 или фиксированная сумма)
    discount_value: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    
    # Минимальная сумма заказа для применения промокода
    min_order_amount: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True)
    
    # Максимальная сумма скидки (для процентных промокодов)
    max_discount_amount: Mapped[Decimal | None] = mapped_column(Numeric(10, 2), nullable=True)
    
    # Количество использований
    max_uses: Mapped[int | None] = mapped_column(Integer, nullable=True)  # None = без ограничений
    uses_count: Mapped[int] = mapped_column(Integer, default=0)
    
    # Лимит использований на одного пользователя
    max_uses_per_user: Mapped[int | None] = mapped_column(Integer, nullable=True, default=1)
    
    # Даты действия
    valid_from: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    valid_until: Mapped[datetime | None] = mapped_column(DateTime, nullable=True)
    
    # Активность
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)
    updated_at: Mapped[datetime] = mapped_column(default=datetime.utcnow, onupdate=datetime.utcnow)

    # Relationships
    business: Mapped["Business"] = relationship("Business", back_populates="promocodes")
    usages: Mapped[list["PromocodeUsage"]] = relationship("PromocodeUsage", back_populates="promocode")


class PromocodeUsage(Base):
    """Модель использования промокода."""

    __tablename__ = "promocode_usages"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    promocode_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("promocodes.id"), nullable=False, index=True)
    order_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("orders.id"), nullable=False, index=True)
    user_telegram_id: Mapped[int | None] = mapped_column(BigInteger, nullable=True, index=True)
    
    # Сумма скидки, применённая к заказу
    discount_amount: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    
    # Сумма заказа до применения промокода
    order_amount_before: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    
    # Сумма заказа после применения промокода
    order_amount_after: Mapped[Decimal] = mapped_column(Numeric(10, 2), nullable=False)
    
    created_at: Mapped[datetime] = mapped_column(default=datetime.utcnow)

    # Relationships
    promocode: Mapped["Promocode"] = relationship("Promocode", back_populates="usages")
    order: Mapped["Order"] = relationship("Order", back_populates="promocode_usages")

File: frontend/lib/mini_app/domain/entities/promocode.dart
import 'package:equatable/equatable.dart';

class Promocode extends Equatable {
  final String id;
  final String businessId;
  final String code;
  final String? description;
  final String discountType;  // "percentage" или "fixed"
  final double discountValue;
  final double? minOrderAmount;
  final double? maxDiscountAmount;
  final int? maxUses;
  final int usesCount;
  final int? maxUsesPerUser;
  final DateTime? validFrom;
  final DateTime? validUntil;
  final bool isActive;
  final DateTime createdAt;
  final DateTime updatedAt;

  const Promocode({
    required this.id,
    required this.businessId,
    required this.code,
    this.description,
    required this.discountType,
    required this.discountValue,
    this.minOrderAmount,
    this.maxDiscountAmount,
    this.maxUses,
    required this.usesCount,
    this.maxUsesPerUser,
    this.validFrom,
    this.validUntil,
    required this.isActive,
    required this.createdAt,
    required this.updatedAt,
  });

  /// Рассчитать размер скидки для указанной суммы заказа
  double calculateDiscount(double orderAmount) {
    double discount;
    
    if (discountType == 'percentage') {
      discount = orderAmount * (discountValue / 100);
      if (maxDiscountAmount != null) {
        discount = discount > maxDiscountAmount! ? maxDiscountAmount! : discount;
      }
    } else {
      discount = discountValue;
    }
    
    // Скидка не может быть больше суммы заказа
    if (discount > orderAmount) {
      discount = orderAmount;
    }
    
    return discount;
  }

  /// Проверить, валиден ли промокод для указанной суммы заказа
  bool isValidForAmount(double orderAmount) {
    if (!isActive) return false;
    
    final now = DateTime.now();
    if (validFrom != null && now.isBefore(validFrom!)) return false;
    if (validUntil != null && now.isAfter(validUntil!)) return false;
    
    if (minOrderAmount != null && orderAmount < minOrderAmount!) return false;
    if (maxUses != null && usesCount >= maxUses!) return false;
    
    return true;
  }

  @override
  List<Object?> get props => [
        id,
        businessId,
        code,
        description,
        discountType,
        discountValue,
        minOrderAmount,
        maxDiscountAmount,
        maxUses,
        usesCount,
        maxUsesPerUser,
        validFrom,
        validUntil,
        isActive,
        createdAt,
        updatedAt,
      ];
}

